Update `diffusers-cli` for agentic use (#13966)
huggingface/diffusers: docs/source/en/using-diffusers/cli.md

> <!--Copyright 2026 The HuggingFace Team. All rights reserved.

Author: @stevhliu Could you give this a look please.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

> # Command line interface (`diffusers-cli`)

Reviewer: ```suggestion
# Command line interface
```

`diffusers-cli` is a command line entry point for running, inspecting, and packaging Diffusers pipelines
without writing Python. It's installed alongside the `diffusers` package.

The CLI adapts its output for whoever's reading it:

- Under an AI coding agent (Claude Code, Cursor, Aider, GitHub Copilot Agent — detected via env vars like
  `CLAUDECODE`, `CURSOR_AI`, `AIDER_AI_CONTEXT`), output switches to a compact `key=value` / TSV format with
  no progress bars.
- In a normal terminal, it prints human-friendly summaries.

> - Force a specific format with `--format {auto, human, agent, json, quiet}` **before** the subcommand:

Reviewer: ```suggestion
- Force a specific format with `--format {auto, human, agent, json, quiet}`
```

  `diffusers-cli --format json run --model ...`.

## Available commands

| Command | Purpose |
|---|---|
| [`env`](#env) | Print environment info for bug reports. |

> | [`schema`](#schema) | Introspect a pipeline's `__call__` signature without downloading weights. |

Reviewer: ```suggestion
| [`schema`](#schema) | Inspect a pipeline's `__call__` signature without downloading weights. |
```

| [`run`](#run) | Run a pipeline locally or submit it to Hugging Face Jobs. |
| [`custom_blocks`](#custom_blocks) | Package a local `ModularPipelineBlocks` subclass for the Hub. |
| [`fp16_safetensors`](#fp16_safetensors) | Convert a checkpoint to fp16 `.safetensors`. |
| [`skills`](#skills) | Install pre-authored skill bundles into your AI coding agent. |

> [!TIP]

> > This page covers the common flags. For the full, always-current list of options for any subcommand, run `diffusers-cli <command> --help` (e.g. `diffusers-cli run --help`).

Reviewer: ```suggestion
> This page covers the common flags. For the full, always-current list of options for any subcommand, run `diffusers-cli <command> --help` (`diffusers-cli run --help`).
```

## `env`

> Prints Python / PyTorch / diffusers versions, CUDA info, and installed optional deps. Use it when opening an

Reviewer: ```suggestion
Prints Python/PyTorch/Diffusers versions, CUDA info, and installed optional deps. Use it when opening an
```

issue so maintainers can reproduce your setup.

```bash
diffusers-cli env
```

## `schema`

Prints the accepted `__call__` arguments for a pipeline, without downloading weights. It fetches the repo's
`model_index.json` (or `modular_model_index.json`), resolves the pipeline class, and inspects its signature.

> For modular repos with custom code, pass `--trust-remote-code`.

Reviewer: ```suggestion
`schema` reads `model_index.json` for standard pipelines, `modular_model_index.json` for modular pipelines, or `modular_config.json` for custom block repositories. Standard pipelines expose their `__call__` signature, and modular repositories expose declared block inputs. For modular repos with custom code, pass `--trust-remote-code`.
```

```bash
diffusers-cli --format json schema --model black-forest-labs/FLUX.1-dev
```

## `run`

Run a pipeline end-to-end. Auto-detects standard vs modular repos, auto-loads media inputs from URLs or local

> paths, saves outputs by sniffing the pipeline's return type, and can submit the same call to Hugging Face Jobs

Reviewer: ```suggestion
paths, saves outputs by detecting the pipeline's return type, and can submit the same call to Hugging Face Jobs
```

via `--remote`.

Minimal example:

```bash
diffusers-cli run \
    --model black-forest-labs/FLUX.1-dev \
    --dtype bf16 \
    --pipeline-kwargs '{"prompt": "an astronaut riding a horse"}'
```

### Passing pipeline arguments

`--pipeline-kwargs` takes a JSON object that's forwarded to `pipeline(**kwargs)`. String values at known
media-input keys are auto-loaded:

> - **Images** (`image`, `mask_image`, `control_image`, `ip_adapter_image`, `image_2`) → `PIL.Image` via

Reviewer: ```suggestion
- Images (`image`, `mask_image`, `control_image`, `ip_adapter_image`, `image_2`) → `PIL.Image` via
```

  [`load_image`].

> - **Videos** (`video`, `control_video`) → `list[PIL.Image]` via [`load_video`].

Reviewer: ```suggestion
- Videos (`video`, `control_video`) → `list[PIL.Image]` via [`load_video`].
```

> - **Audio** (`initial_audio_waveforms`, `reference_audio`, `src_audio`) → `torch.Tensor` via `torchaudio.load`.

Reviewer: ```suggestion
- Audio (`initial_audio_waveforms`, `reference_audio`, `src_audio`) → `torch.Tensor` via `torchaudio.load`.
```

```bash
diffusers-cli run \
    --model black-forest-labs/FLUX.2-klein-9B --dtype bf16 \
    --pipeline-kwargs '{"prompt": "make the fur grey", "image": "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/cat.png", "strength": 0.6}'
```

> ### Loading

Reviewer: ```suggestion
### Loading

Configure how the CLI loads model weights and custom pipeline code.
```

- `--dtype {auto, bf16, fp16, fp32, ...}` — weight dtype. See

>   [Pipeline data types](../using-diffusers/loading#pipeline-data-types).

Reviewer: ```suggestion
- `--dtype {auto, bfloat16, bf16, float16, fp16, float32, fp32}` — weight dtype.
```

  `balanced` (auto-splits components across visible GPUs), or a JSON dict for explicit per-component placement.
  Auto-detected if omitted. See [device_map](../training/distributed_inference#device_map) for more details
- `--variant fp16` — pick a weight variant.
- `--revision <sha>` — pin a specific model revision.
- `--trust-remote-code` — allow custom code from the Hub (required for repos that ship custom pipeline classes
  or modular blocks). See [Community pipelines](../using-diffusers/custom_pipeline_overview) for standard custom
  pipelines and [Modular Diffusers](../modular_diffusers/overview).
- `--lora <spec>` — attach one or more LoRA adapters after loading. Accepts a single JSON object
  or a list of them. `lora_id` is required per entry; `lora_scale` defaults to `1.0`; `adapter_name`
  is optional (auto-generated as `lora_<i>` when stacking).
  - Single: `--lora '{"lora_id": "alvdansen/flux-koda", "lora_scale": 0.8}'`
  - Multiple: `--lora '[{"lora_id": "alvdansen/flux-koda", "lora_scale": 0.6, "adapter_name": "koda"}, {"lora_id": "Shakker-Labs/FLUX.1-dev-LoRA-AntiBlur", "lora_scale": 0.4}]'`

>   All specs are loaded via `pipeline.load_lora_weights(...)`, then activated together via a single

Reviewer: ```suggestion
  All specs are loaded by `pipeline.load_lora_weights(...)`, then activated together with a single
```

  `enable_group_offload(offload_type="leaf_level", use_stream=True)`. Onload target comes from `--device-map`
  (which must be a plain device string for offload). See
  [Model offloading](../optimization/memory#model-offloading) and
  [Group offloading](../optimization/memory#group-offloading).
- `--attention-backend {default, flash_hub, flash_varlen_hub, flash_4_hub, sage_hub}` — Hub-hosted attention
  kernels, auto-downloaded on first use. Transformer-based pipelines only; ignored with a warning on legacy UNet
  pipelines. See [Attention backends](../optimization/attention_backends).
- `--vae-tiling` / `--vae-slicing` — lower VAE decode VRAM. See
  [VAE tiling](../optimization/memory#vae-tiling) and [VAE slicing](../optimization/memory#vae-slicing).
- `--compile [JSON]` — `torch.compile` every denoiser submodule. Prefers regional compilation via
  `compile_repeated_blocks` where available. Bare `--compile` uses `fullgraph=true`; a JSON object is forwarded
  to `torch.compile`. Currently not supported with `--context-parallel` — the CLI logs a warning and skips the
  compile step in that case. See [torch.compile](../optimization/fp16#torchcompile) and

>   [Regional compilation](../optimization/fp16#regional-compilation).

Reviewer: ```suggestion
- `--compile [JSON]` — compile denoiser modules with [torch.compile](../optimization/fp16#torchcompile). The CLI prefers [regional compilation](../optimization/fp16#regional-compilation) for modules with repeated blocks. Bare `--compile` uses `fullgraph=true`. A JSON object is forwarded to `torch.compile`. Not supported with `--context-parallel`.
```

- `--context-parallel` — Ulysses-style context parallelism on a DiT-based pipeline. Locally requires torchrun;
  under `--remote` the CLI wraps `torchrun --nproc-per-node=gpu` for you. See
  [Context parallelism](../training/distributed_inference#context-parallelism).

### Outputs

> `run` sniffs the pipeline output type:

Reviewer: ```suggestion
`run` detects the pipeline output type:
```

> - `PIL.Image` / list → `<task>-<i>.png`

Reviewer: ```suggestion
- `PIL.Image`/list → `<task>-<i>.png`
```

- Frame sequence → `<task>-0.mp4` (`--fps` controls framerate, default 8)
- Audio array → `<task>-0.wav` (`--sampling-rate` controls rate)
- Anything else → JSON dump

Default output directory is `~/.diffusers/cli/run/outputs/diffusers-run-<YYYYMMDDTHHMMSS>-<uuid>/`. Each run
gets its own subdirectory so consecutive invocations don't overwrite. Override with `--output <path>` (file or

> directory) — explicit paths bypass the namespaced default and land flat.

Reviewer: ```suggestion
directory) — explicit paths bypass the namespaced default.
```

Use `--push-to <your-bucket-id>` to upload outputs to a
[Hugging Face storage bucket](https://huggingface.co/docs/hub/en/storage-buckets). `<your-bucket-id>` is a
bare bucket id in `<namespace>/<name>` form (same shape as a Hub repo id) — `hf://` URIs are not accepted here.
The bucket is created if missing; objects land under `<run_id>/<filename>` and are addressable as
`hf://buckets/<your-bucket-id>/<run_id>/<filename>`.

> 

Reviewer: ```suggestion
The table below describes remote runs. For local runs, `--push-to` uploads the locally saved output. It does not suppress local file creation.

```

| `--push-to` set? | `--output` set? | Result |
|---|---|---|
| no | no | download to default local dir |
| no | yes | download to `--output` |
| yes | no | bucket only, no local download |
| yes | yes | bucket AND `--output` |

> `--format` shapes the **stdout metadata** (paths, timing, job info) — it does not change the file format of

Reviewer: ```suggestion
`--format` shapes the stdout metadata (paths, timing, job info) — it does not change the file format of
```

config from the pipeline `state` at call time.

## `fp16_safetensors`

Convert a checkpoint on the Hub to fp16 `.safetensors` and push the result. Useful for shrinking a repo's
weight size for faster loading. See `diffusers-cli fp16_safetensors --help` for the exact args.

## `skills`

Install Agent Skills bundles from the diffusers repo (`.ai/skills/`).

```bash
# Install the diffusers-cli skill
diffusers-cli skills add diffusers-cli

# Install every skill in the registry
diffusers-cli skills add --all

# List available skills
diffusers-cli skills list

# Preview a skill's SKILL.md without installing
diffusers-cli skills preview diffusers-cli

# Refetch and reinstall every managed skill
diffusers-cli skills update

# Install to the user-level directory instead of the current project
diffusers-cli skills add diffusers-cli --global
```

`--force` overwrites an existing install. `--all` fetches every skill in the registry in one call
and downgrades individual failures to warnings so one broken skill doesn't abort the batch.

<!-- Doc link references -->
[`load_image`]: /docs/diffusers/api/utilities#diffusers.utils.load_image

> [`load_video`]: /docs/diffusers/api/utilities#diffusers.utils.load_video

Reviewer: i think these can be removed?
